sweep/sweeper.go

```diff
@@ -0,0 +1,234 @@
+	// function to generate the final witness required for spending.
+	addWitness := func(idx int, tso Input) error {
+		witness, err := tso.BuildWitness(
+			s.cfg.Signer, sweepTx, hashCache, idx,
+		)
+		if err != nil {
+			return err
+		}
+
+		sweepTx.TxIn[idx].Witness = witness
+
+		return nil
+	}
+
+	// Finally we'll attach a valid witness to each csv and cltv input
+	// within the sweeping transaction.
+	for i, input := range inputs {
+		if err := addWitness(i, input); err != nil {
+			return nil, err
+		}
+	}
+
+	return sweepTx, nil
+}
+
+// getWeightEstimate returns a weight estimate for the given inputs.
+// Additionally, it returns counts for the number of csv and cltv inputs.
+func (s *UtxoSweeper) getWeightEstimate(inputs []Input) (int64, int, int) {
+
+	// Create a transaction which sweeps all the newly mature outputs into
+	// an output controlled by the wallet.
+
+	// TODO(roasbeef): can be more intelligent about buffering outputs to
+	// be more efficient on-chain.
+
+	var weightEstimate lnwallet.TxWeightEstimator
+
+	// Our sweep transaction will pay to a single segwit p2wkh address,
+	// ensure it contributes to our weight estimate.
+	weightEstimate.AddP2WKHOutput()
+
+	// For each output, use its witness type to determine the estimate
+	// weight of its witness, and add it to the proper set of spendable
+	// outputs.
+	csvCount := 0
+	cltvCount := 0
+
+	for i := range inputs {
+		input := inputs[i]
+
+		switch input.WitnessType() {
+
+		// Outputs on a remote commitment transaction that pay directly
+		// to us.
+		case lnwallet.CommitmentNoDelay:
+			weightEstimate.AddP2WKHInput()
+
+		// Outputs on a past commitment transaction that pay directly
+		// to us.
+		case lnwallet.CommitmentTimeLock:
+			weightEstimate.AddWitnessInput(
+				lnwallet.ToLocalTimeoutWitnessSize,
+			)
+			csvCount++
+
+		// Outgoing second layer HTLC's that have confirmed within the
+		// chain, and the output they produced is now mature enough to
+		// sweep.
+		case lnwallet.HtlcOfferedTimeoutSecondLevel:
+			weightEstimate.AddWitnessInput(
+				lnwallet.ToLocalTimeoutWitnessSize,
+			)
+			csvCount++
+
+		// Incoming second layer HTLC's that have confirmed within the
+		// chain, and the output they produced is now mature enough to
+		// sweep.
+		case lnwallet.HtlcAcceptedSuccessSecondLevel:
+			weightEstimate.AddWitnessInput(
+				lnwallet.ToLocalTimeoutWitnessSize,
+			)
+			csvCount++
+
+		// An HTLC on the commitment transaction of the remote party,
+		// that has had its absolute timelock expire.
+		case lnwallet.HtlcOfferedRemoteTimeout:
+			weightEstimate.AddWitnessInput(
+				lnwallet.AcceptedHtlcTimeoutWitnessSize,
+			)
+			cltvCount++
+
+		// An HTLC on the commitment transaction of the remote party,
+		// that can be swept with the preimage.
+		case lnwallet.HtlcAcceptedRemoteSuccess:
+			weightEstimate.AddWitnessInput(
+				lnwallet.OfferedHtlcSuccessWitnessSize,
+			)
+
+		default:
+			log.Warnf("kindergarten output in nursery store "+
```

Comment:
i think this is consistent with the current behavior, but currently we won't add any weight estimate for unknown witness types

Reply:
i take that back, the prior version would properly filter unknown output types and not include them in the transaction. the new version will include them